// @terra-money/wallet-controller/operators/getExtensions.js
import { of, switchMap } from 'rxjs';
import { fromFetch } from 'rxjs/fetch';
import { catchError, map } from 'rxjs/operators';
import { getDesktopBrowserType } from '../utils/browser-check';
const FALLBACK = {
    whitelist: [
        {
            name: 'Terra Station',
            identifier: 'station',
            icon: 'https://assets.terra.money/icon/station-extension/icon.png',
            urls: [
                {
                    browser: 'chrome',
                    url: 'https://chrome.google.com/webstore/detail/terra-station/aiifbnbfobpmeekipheeijimdpnlpgpp',
                },
            ],
        },
        {
            name: 'XDEFI Wallet',
            identifier: 'xdefi-wallet',
            icon: 'https://xdefi-prod-common-ui.s3.eu-west-1.amazonaws.com/logo.svg',
            urls: [
                {
                    browser: 'chrome',
                    url: 'https://chrome.google.com/webstore/detail/xdefi-wallet/hmeobnfnfcmdkdcmlblgagmfpfboieaf',
                },
            ],
        },
    ],
};
export function getExtensions() {
    const currentBrowser = getDesktopBrowserType(navigator.userAgent);
    if (!currentBrowser) {
        return of([]);
    }
    return fromFetch('https://assets.terra.money/extensions.json').pipe(switchMap((res) => {
        if (res.ok) {
            return res.json();
        }
        else {
            return of(FALLBACK);
        }
    }), catchError(() => {
        return of(FALLBACK);
    }), map(({ whitelist }) => {
        return whitelist
            .filter(({ urls }) => urls.some(({ browser }) => currentBrowser === browser))
            .map(({ name, identifier, icon, urls }) => {
            return {
                name,
                identifier,
                icon,
                url: urls.find(({ browser }) => currentBrowser === browser).url,
            };
        });
    }));
}
